### ISSUE TITLE
undefined method forem_needs_moderation?

### ISSUE THREAD

**Issue by nazrulcse** [rubysherpas/forem#415]
```
C:\Users\Nazrul\ugl>rails g forem:install
Copying over Forem migrations...
What is your user class called? [User]
What is the current_user helper called in your app? [current_user]
Defining forem_user method inside ApplicationController...
      insert  app/controllers/application_controller.rb
Adding forem initializer (config/initializers/forem.rb)...
      create  config/initializers/forem.rb
Running rake db:migrate
Creating default forum and topic
C:/Ruby193/lib/ruby/gems/1.9.1/gems/activemodel-3.2.13/lib/active_model/attribute_methods.rb:407:
in `method_missing': undefined method `forem_needs_moderation?' for #<User:0x51b7a98> (NoMethodError)
      from C:/Ruby193/lib/ruby/gems/1.9.1/gems/activerecord-3.2.13/lib/active_record/attribute_methods.rb:149:in `method_missing'
      from C:/Ruby193/lib/ruby/gems/1.9.1/gems/activesupport-3.2.13/lib/active_support/core_ext/object/try.rb:36:in `try'
```

**Comment by parndt** [comment]
This is confusing - is there a way that we can access your application (like, can you open source it) to aid us in reproducing this error?

**Comment by oleander** [comment]
I get the same error.

I'm using 7434b1d778b4350e1ad24acf78a5051155576637 with rails 3.2.11 and ruby 1.9.3.

```
  rails g forem:install                                                                                                                            [linus-oleander.lan][18:12:15]
Copying over Forem migrations...
What is your user class called? [User]
What is the current_user helper called in your app? [current_user]
Defining forem_user method inside ApplicationController...
      insert  app/controllers/application_controller.rb
Adding forem initializer (config/initializers/forem.rb)...
      create  config/initializers/forem.rb
Running rake db:migrate
NOTICE:  CREATE TABLE will create implicit sequence "forem_forums_id_seq" for serial column "forem_forums.id"
NOTICE:  CREATE TABLE / PRIMARY KEY will create implicit index "forem_forums_pkey" for table "forem_forums"
NOTICE:  CREATE TABLE will create implicit sequence "forem_topics_id_seq" for serial column "forem_topics.id"
NOTICE:  CREATE TABLE / PRIMARY KEY will create implicit index "forem_topics_pkey" for table "forem_topics"
NOTICE:  CREATE TABLE will create implicit sequence "forem_posts_id_seq" for serial column "forem_posts.id"
NOTICE:  CREATE TABLE / PRIMARY KEY will create implicit index "forem_posts_pkey" for table "forem_posts"
NOTICE:  CREATE TABLE will create implicit sequence "forem_views_id_seq" for serial column "forem_views.id"
NOTICE:  CREATE TABLE / PRIMARY KEY will create implicit index "forem_views_pkey" for table "forem_views"
NOTICE:  CREATE TABLE will create implicit sequence "forem_categories_id_seq" for serial column "forem_categories.id"
NOTICE:  CREATE TABLE / PRIMARY KEY will create implicit index "forem_categories_pkey" for table "forem_categories"
NOTICE:  CREATE TABLE will create implicit sequence "forem_subscriptions_id_seq" for serial column "forem_subscriptions.id"
NOTICE:  CREATE TABLE / PRIMARY KEY will create implicit index "forem_subscriptions_pkey" for table "forem_subscriptions"
NOTICE:  CREATE TABLE will create implicit sequence "forem_groups_id_seq" for serial column "forem_groups.id"
NOTICE:  CREATE TABLE / PRIMARY KEY will create implicit index "forem_groups_pkey" for table "forem_groups"
NOTICE:  CREATE TABLE will create implicit sequence "forem_memberships_id_seq" for serial column "forem_memberships.id"
NOTICE:  CREATE TABLE / PRIMARY KEY will create implicit index "forem_memberships_pkey" for table "forem_memberships"
NOTICE:  CREATE TABLE will create implicit sequence "forem_moderator_groups_id_seq" for serial column "forem_moderator_groups.id"
NOTICE:  CREATE TABLE / PRIMARY KEY will create implicit index "forem_moderator_groups_pkey" for table "forem_moderator_groups"
Creating default forum and topic
/Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/activemodel-3.2.11/lib/active_model/attribute_methods.rb:407:in `method_missing': undefined method `forem_needs_moderation?' for #<User:0x007ffe08faea70> (NoMethodError)
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/activerecord-3.2.11/lib/active_record/attribute_methods.rb:149:in `method_missing'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/activesupport-3.2.11/lib/active_support/core_ext/object/try.rb:36:in `try'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/bundler/gems/forem-7434b1d778b4/app/models/forem/post.rb:116:in `skip_pending_review'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/activesupport-3.2.11/lib/active_support/callbacks.rb:417:in `_run__3755572121445268244__create__3984414227987606957__callbacks'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/activesupport-3.2.11/lib/active_support/callbacks.rb:405:in `__run_callback'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/activesupport-3.2.11/lib/active_support/callbacks.rb:385:in `_run_create_callbacks'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/activesupport-3.2.11/lib/active_support/callbacks.rb:81:in `run_callbacks'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/activerecord-3.2.11/lib/active_record/callbacks.rb:268:in `create'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/activerecord-3.2.11/lib/active_record/persistence.rb:348:in `create_or_update'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/activerecord-3.2.11/lib/active_record/callbacks.rb:264:in `block in create_or_update'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/activesupport-3.2.11/lib/active_support/callbacks.rb:447:in `_run__3755572121445268244__save__3984414227987606957__callbacks'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/activesupport-3.2.11/lib/active_support/callbacks.rb:405:in `__run_callback'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/activesupport-3.2.11/lib/active_support/callbacks.rb:385:in `_run_save_callbacks'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/activesupport-3.2.11/lib/active_support/callbacks.rb:81:in `run_callbacks'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/activerecord-3.2.11/lib/active_record/callbacks.rb:264:in `create_or_update'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/activerecord-3.2.11/lib/active_record/persistence.rb:84:in `save'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/activerecord-3.2.11/lib/active_record/validations.rb:50:in `save'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/activerecord-3.2.11/lib/active_record/attribute_methods/dirty.rb:22:in `save'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/activerecord-3.2.11/lib/active_record/transactions.rb:259:in `block (2 levels) in save'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/activerecord-3.2.11/lib/active_record/transactions.rb:313:in `block in with_transaction_returning_status'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/activerecord-3.2.11/lib/active_record/connection_adapters/abstract/database_statements.rb:192:in `transaction'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/activerecord-3.2.11/lib/active_record/transactions.rb:208:in `transaction'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/newrelic_rpm-3.5.6.55/lib/new_relic/agent/method_tracer.rb:491:in `block in transaction_with_trace_ActiveRecord_self_name_transaction'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/newrelic_rpm-3.5.6.55/lib/new_relic/agent/method_tracer.rb:240:in `trace_execution_scoped'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/newrelic_rpm-3.5.6.55/lib/new_relic/agent/method_tracer.rb:486:in `transaction_with_trace_ActiveRecord_self_name_transaction'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/activerecord-3.2.11/lib/active_record/transactions.rb:311:in `with_transaction_returning_status'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/activerecord-3.2.11/lib/active_record/transactions.rb:259:in `block in save'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/activerecord-3.2.11/lib/active_record/transactions.rb:270:in `rollback_active_record_state!'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/activerecord-3.2.11/lib/active_record/transactions.rb:258:in `save'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/activerecord-3.2.11/lib/active_record/associations/has_many_association.rb:16:in `insert_record'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/activerecord-3.2.11/lib/active_record/autosave_association.rb:346:in `block in save_collection_association'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/activerecord-3.2.11/lib/active_record/autosave_association.rb:337:in `each'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/activerecord-3.2.11/lib/active_record/autosave_association.rb:337:in `save_collection_association'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/activerecord-3.2.11/lib/active_record/autosave_association.rb:191:in `block in add_autosave_association_callbacks'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/activerecord-3.2.11/lib/active_record/autosave_association.rb:161:in `instance_eval'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/activerecord-3.2.11/lib/active_record/autosave_association.rb:161:in `block in define_non_cyclic_method'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/activesupport-3.2.11/lib/active_support/callbacks.rb:413:in `_run__1824031267506073167__create__3984414227987606957__callbacks'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/activesupport-3.2.11/lib/active_support/callbacks.rb:405:in `__run_callback'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/activesupport-3.2.11/lib/active_support/callbacks.rb:385:in `_run_create_callbacks'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/activesupport-3.2.11/lib/active_support/callbacks.rb:81:in `run_callbacks'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/activerecord-3.2.11/lib/active_record/callbacks.rb:268:in `create'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/activerecord-3.2.11/lib/active_record/persistence.rb:348:in `create_or_update'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/activerecord-3.2.11/lib/active_record/callbacks.rb:264:in `block in create_or_update'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/activesupport-3.2.11/lib/active_support/callbacks.rb:458:in `_run__1824031267506073167__save__3984414227987606957__callbacks'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/activesupport-3.2.11/lib/active_support/callbacks.rb:405:in `__run_callback'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/activesupport-3.2.11/lib/active_support/callbacks.rb:385:in `_run_save_callbacks'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/activesupport-3.2.11/lib/active_support/callbacks.rb:81:in `run_callbacks'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/activerecord-3.2.11/lib/active_record/callbacks.rb:264:in `create_or_update'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/activerecord-3.2.11/lib/active_record/persistence.rb:104:in `save!'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/activerecord-3.2.11/lib/active_record/validations.rb:56:in `save!'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/activerecord-3.2.11/lib/active_record/attribute_methods/dirty.rb:33:in `save!'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/activerecord-3.2.11/lib/active_record/transactions.rb:264:in `block in save!'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/activerecord-3.2.11/lib/active_record/transactions.rb:313:in `block in with_transaction_returning_status'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/activerecord-3.2.11/lib/active_record/connection_adapters/abstract/database_statements.rb:192:in `transaction'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/activerecord-3.2.11/lib/active_record/transactions.rb:208:in `transaction'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/newrelic_rpm-3.5.6.55/lib/new_relic/agent/method_tracer.rb:491:in `block in transaction_with_trace_ActiveRecord_self_name_transaction'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/newrelic_rpm-3.5.6.55/lib/new_relic/agent/method_tracer.rb:240:in `trace_execution_scoped'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/newrelic_rpm-3.5.6.55/lib/new_relic/agent/method_tracer.rb:486:in `transaction_with_trace_ActiveRecord_self_name_transaction'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/activerecord-3.2.11/lib/active_record/transactions.rb:311:in `with_transaction_returning_status'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/activerecord-3.2.11/lib/active_record/transactions.rb:264:in `save!'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/bundler/gems/forem-7434b1d778b4/db/seeds.rb:17:in `<top (required)>'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/zeus-0.13.3/lib/zeus/load_tracking.rb:50:in `load'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/zeus-0.13.3/lib/zeus/load_tracking.rb:50:in `load'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/zeus-0.13.3/lib/zeus/load_tracking.rb:43:in `load'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/activesupport-3.2.11/lib/active_support/dependencies.rb:245:in `block in load'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/activesupport-3.2.11/lib/active_support/dependencies.rb:236:in `load_dependency'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/activesupport-3.2.11/lib/active_support/dependencies.rb:245:in `load'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/railties-3.2.11/lib/rails/engine.rb:520:in `load_seed'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/railties-3.2.11/lib/rails/railtie/configurable.rb:30:in `method_missing'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/bundler/gems/forem-7434b1d778b4/lib/generators/forem/install_generator.rb:67:in `seed_database'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/thor-0.17.0/lib/thor/task.rb:27:in `run'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/thor-0.17.0/lib/thor/invocation.rb:120:in `invoke_task'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/thor-0.17.0/lib/thor/invocation.rb:126:in `block in invoke_all'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/thor-0.17.0/lib/thor/invocation.rb:126:in `each'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/thor-0.17.0/lib/thor/invocation.rb:126:in `map'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/thor-0.17.0/lib/thor/invocation.rb:126:in `invoke_all'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/thor-0.17.0/lib/thor/group.rb:238:in `dispatch'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/thor-0.17.0/lib/thor/base.rb:434:in `start'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/railties-3.2.11/lib/rails/generators.rb:171:in `invoke'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/railties-3.2.11/lib/rails/commands/generate.rb:12:in `<top (required)>'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/activesupport-3.2.11/lib/active_support/dependencies.rb:251:in `require'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/activesupport-3.2.11/lib/active_support/dependencies.rb:251:in `block in require'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/activesupport-3.2.11/lib/active_support/dependencies.rb:236:in `load_dependency'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/activesupport-3.2.11/lib/active_support/dependencies.rb:251:in `require'
  from /Users/linus/.rvm/gems/ruby-1.9.3-p392/gems/railties-3.2.11/lib/rails/commands.rb:29:in `<top (required)>'
  from script/rails:6:in `require'
  from script/rails:6:in `<main>'
rails g forem:install  51,02s user 9,53s system 52% cpu 1:54,67 total
```

**Comment by oleander** [comment]
I did some research on the subject and I think I solved it.

When running `rails g forem:install` a init file is created, **but not loaded** (which is the core problem). In this init file `Forem.user_class` is being set. Later in installation `Post#skip_pending_review` is being called which runs the following line `user.try(:forem_needs_moderation?)`. The problem is that the `User` class has not been infected/injected using `Forem.user_class.class_eval` so the `User` class doesn't respond to `forem_needs_moderation?` . This means that  the error above will be raised. 

tl;dr The solution is to run the `rails g forem:install` twice.

**Comment by parndt** [comment]
@oleander thanks for this. Do you think the following solution might work?

``` diff
diff --git a/lib/generators/forem/install_generator.rb b/lib/generators/forem/install_generator.rb
index 69296a9..49e2d96 100644
--- a/lib/generators/forem/install_generator.rb
+++ b/lib/generators/forem/install_generator.rb
@@ -50,6 +50,7 @@ module Forem
         else
           puts "Adding forem initializer (config/initializers/forem.rb)..."
           template "initializer.rb", path
+          require path # Load the configuration per issue #415
         end
       end
```

**Comment by oleander** [comment]
Yes, that should solve the problem. I haven't tried it, but it looks like it should work.

**Comment by parndt** [comment]
The installer still works fine with the line in there so I think we should give it a go!

**Comment by parndt** [comment]
@oleander I pushed 7447c7e which is that diff. If you can still reproduce the issue could you `bundle update --source forem` and give it a go please? You may have to delete `config/initializers/forem.rb` or something.
